### ISSUE TITLE
Azure Key Vault not being successfully mounted on pod

### ISSUE THREAD

**Issue by JuanFryS** [Azure/secrets-store-csi-driver-provider-azure#292]
**What steps did you take and what happened:**
Followed MS guide to mount Azure Key Vault Secrets into pods: [MS Guide](https://docs.microsoft.com/en-us/azure/key-vault/general/key-vault-integrate-kubernetes).

- Created Service Principal and key vault instance and granted permissions for SP to access the key vault and retrieve secrets

- Installed the Secrets Store CSI driver and the Azure Key Vault provider for the driver with:

```
helm repo add csi-secrets-store-provider-azure https://raw.githubusercontent.com/Azure/secrets-store-csi-driver-provider-azure/master/charts
helm install csi-secrets-store-provider-azure/csi-secrets-store-provider-azure --generate-name
```

- Created the following SecretProviderClass object:

```
apiVersion: secrets-store.csi.x-k8s.io/v1alpha1
kind: SecretProviderClass
metadata:
  name: "keyvault"
spec:
  provider: azure
  parameters:
    keyvaultName: "dev1-keyvault"
    objects:  |
      array:
        - |
          objectName: example
          objectType: secret
          objectVersion: ""
        - |
          objectName: secret1
          objectType: secret
          objectVersion: ""
    tenantId: "X-X-X-X-X"
```
- Deployed following single pod where secrets should be mounted:
```
kind: Pod
apiVersion: v1
metadata:
  name: busybox-keyvault
spec:
  imagePullSecrets:
   - name: <credentials_repo>
  containers:
   - name: busybox
     image: <private_repo>
     command:
       - sleep
       - "3600"
     volumeMounts:
      - name: secrets
        mountPath: "/mnt/secrets"
        readOnly: true
  volumes:
    - name: secrets
      csi:
        driver: secrets-store.csi.k8s.io
        readOnly: true
        volumeAttributes:
          secretProviderClass: "keyvault"
        nodePublishSecretRef:
          name: secrets-store-creds
```

**What did you expect to happen:**
Secrets mounted into pod under the '/mnt/secrets' folder

**Anything else you would like to add:**
Pod seems stuck into 'ContainerCreating' Status.
By describing the pod following messages can be seen:

```
Events:
  Type     Reason       Age    From               Message
  ----     ------       ----   ----               -------
 Normal   Scheduled    8m7s                default-scheduler  Successfully assigned keyvault-test/busybox-keyvault to aks-workflow-69744767-vmss000001
  Warning  FailedMount  91s (x3 over 6m4s)  kubelet            Unable to attach or mount volumes: unmounted volumes=[secrets], unattached volumes=[secrets default-token-lp5kh]: timed out waiting for the condition
  Warning  FailedMount  3s (x4 over 6m7s)   kubelet            MountVolume.SetUp failed for volume "secrets" : kubernetes.io/csi: mounter.SetupAt failed: rpc error: code = DeadlineExceeded desc = context deadline exceeded
```

- A ticket was created directly with the Azure technical support from Microsoft. Given this is an open source project not supported by them, we were advised to create an issue.
- Cluster was created after March 2020.
- Most of the time, after hours, they eventually get created and secrets are mounted. Sometimes it can take 2 hours, sometimes 12 and, sometimes, even 24 hours later, secrets are not mounted.

**Which access mode did you use to access the Azure Key Vault instance:**
Service Principal

**Environment:**

- Kubernetes version: v1.16.10
- Cluster type: AKS

**Comment by aramase** [comment]
Could you provider more info - 

1. Which version of the driver and provider are running on the cluster?
2. Logs from the driver pods running on the same node as the workload
3. Logs from providers pods running on the same node as the workload
4. `kubectl get nodes -o wide` output

**Comment by JuanFryS** [comment]
Thanks for the response.
**1. Which version of the driver and provider are running on the cluster?**
Version of the driver is 0.0.14
Version of the provider is 0.0.9
**2. Logs from the driver pods running on the same node as the workload**
```
I1019 16:19:01.825797       1 main.go:110] Version: v1.2.0-0-g6ef000ae
I1019 16:19:01.825841       1 main.go:120] Attempting to open a gRPC connection with: "/csi/csi.sock"
I1019 16:19:01.825851       1 connection.go:151] Connecting to unix:///csi/csi.sock
I1019 16:19:03.898564       1 main.go:127] Calling CSI driver to discover driver name
I1019 16:19:03.898644       1 connection.go:180] GRPC call: /csi.v1.Identity/GetPluginInfo
I1019 16:19:03.898681       1 connection.go:181] GRPC request: {}
I1019 16:19:03.922548       1 connection.go:183] GRPC response: {"name":"secrets-store.csi.k8s.io","vendor_version":"0.0.13"}
I1019 16:19:03.922993       1 connection.go:184] GRPC error: <nil>
I1019 16:19:03.922999       1 main.go:137] CSI driver name: "secrets-store.csi.k8s.io"
I1019 16:19:03.923050       1 node_register.go:58] Starting Registration Server at: /registration/secrets-store.csi.k8s.io-reg.sock
I1019 16:19:03.923192       1 node_register.go:67] Registration Server started at: /registration/secrets-store.csi.k8s.io-reg.sock
I1019 16:19:03.954512       1 main.go:77] Received GetInfo call: &InfoRequest{}
I1019 16:19:04.954576       1 main.go:77] Received GetInfo call: &InfoRequest{}
I1019 16:19:04.990749       1 main.go:87] Received NotifyRegistrationStatus call: &RegistrationStatus{PluginRegistered:true,Error:,}
```
**3. Logs from providers pods running on the same node as the workload**
```
E1029 17:33:40.434585       1 grpc.go:29] GRPC error: failed to mount objects, error: failed to get objectType:secret, objectName:example, objectVersion:: keyvault.BaseClient#GetSecret: Failure sending request: StatusCode=0 -- Original Error: context deadline exceeded
I1029 17:35:42.482023       1 provider.go:243] unmarshaled keyVaultObjects: [{example   secret  } {secret1   secret  }]
I1029 17:35:42.482431       1 provider.go:244] keyVaultObjects len: 2
I1029 17:35:42.483008       1 auth.go:151] azure: using client_id: 3f4f##### REDACTED #####43c3, client_secret: uX6u##### REDACTED #####gQ4~ to retrieve access token
E1029 17:37:42.461313       1 server.go:54] failed to process mount request, error: keyvault.BaseClient#GetSecret: Failure sending request: StatusCode=0 -- Original Error: context deadline exceeded
failed to get objectType:secret, objectName:example, objectVersion:
github.com/Azure/secrets-store-csi-driver-provider-azure/pkg/provider.wrapObjectTypeError
        /home/azureuser/agent/_work/24/s/go/src/github.com/Azure/secrets-store-csi-driver-provider-azure/pkg/provider/provider.go:423
github.com/Azure/secrets-store-csi-driver-provider-azure/pkg/provider.(*Provider).GetKeyVaultObjectContent
        /home/azureuser/agent/_work/24/s/go/src/github.com/Azure/secrets-store-csi-driver-provider-azure/pkg/provider/provider.go:305
github.com/Azure/secrets-store-csi-driver-provider-azure/pkg/provider.(*Provider).MountSecretsStoreObjectContent
        /home/azureuser/agent/_work/24/s/go/src/github.com/Azure/secrets-store-csi-driver-provider-azure/pkg/provider/provider.go:262
github.com/Azure/secrets-store-csi-driver-provider-azure/pkg/server.(*CSIDriverProviderServer).Mount
        /home/azureuser/agent/_work/24/s/go/src/github.com/Azure/secrets-store-csi-driver-provider-azure/pkg/server/server.go:52
sigs.k8s.io/secrets-store-csi-driver/provider/v1alpha1._CSIDriverProvider_Mount_Handler.func1
        /home/azureuser/agent/_work/24/s/go/pkg/mod/sigs.k8s.io/secrets-store-csi-driver@v0.0.14/provider/v1alpha1/service.pb.go:687
github.com/Azure/secrets-store-csi-driver-provider-azure/pkg/utils.LogGRPC
        /home/azureuser/agent/_work/24/s/go/src/github.com/Azure/secrets-store-csi-driver-provider-azure/pkg/utils/grpc.go:27
sigs.k8s.io/secrets-store-csi-driver/provider/v1alpha1._CSIDriverProvider_Mount_Handler
        /home/azureuser/agent/_work/24/s/go/pkg/mod/sigs.k8s.io/secrets-store-csi-driver@v0.0.14/provider/v1alpha1/service.pb.go:689
google.golang.org/grpc.(*Server).processUnaryRPC
        /home/azureuser/agent/_work/24/s/go/pkg/mod/google.golang.org/grpc@v1.31.0/server.go:1180
google.golang.org/grpc.(*Server).handleStream
        /home/azureuser/agent/_work/24/s/go/pkg/mod/google.golang.org/grpc@v1.31.0/server.go:1503
google.golang.org/grpc.(*Server).serveStreams.func1.2
        /home/azureuser/agent/_work/24/s/go/pkg/mod/google.golang.org/grpc@v1.31.0/server.go:843
runtime.goexit
        /usr/lib/go-1.15/src/runtime/asm_amd64.s:1374
```
**4. `kubectl get nodes -o wide` output**
```
NAME                                STATUS   ROLES   AGE   VERSION    INTERNAL-IP   EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION      CONTAINER-RUNTIME
aks-workflow-69744767-vmss000000    Ready    agent   70d   v1.16.10   172.26.5.4    <none>        Ubuntu 16.04.6 LTS   4.15.0-1093-azure   docker://3.0.10+azure
aks-workflow-69744767-vmss000001    Ready    agent   70d   v1.16.10   172.26.5.5    <none>        Ubuntu 16.04.6 LTS   4.15.0-1095-azure   docker://3.0.10+azure
aks-workflow-69744767-vmss000002    Ready    agent   70d   v1.16.10   172.26.5.6    <none>        Ubuntu 16.04.6 LTS   4.15.0-1092-azure   docker://3.0.10+azure```

Please do let me know if any other log is required.

**Comment by aramase** [comment]
Thank you for the logs.

```
E1029 17:37:42.461313       1 server.go:54] failed to process mount request, error: keyvault.BaseClient#GetSecret: Failure sending request: StatusCode=0 -- Original Error: context deadline exceeded
```

This seems to be a timeout error while requesting for a token from AAD and using that to access Key Vault. 
1. Do you have any firewall rules blocking egress?
2. Network policies configured on the cluster that's blocking traffic? 
3. Are you able to get on the node and check if there is connectivity to AAD from the node?

**Comment by JuanFryS** [comment]
Thank you for the investigation. We weren't able to understand that was the problem from the logs.

We have been doing tests today and we have found out the problem was the firewall blocking the outgoing traffic to the key vault.

Working now. Thanks

**Comment by ramooncamacho** [comment]
I've had the same problem using AKS. I realized the kubernetes cluster vnet was not configured to use Azure-provided DNS server, but Google and Microsoft defined mannually. Because of that, some requests to Azure KeyVault was done via internet, and others via private endpoint. That's why intermittent. I forced Azure-provided DNS, to get only private endpoint, and it was solved. I was not expecting this communication through internet, and probably it was blocking as mentioned by @JuanFryS. Just adding this information to help others.
